### PR TITLE
feat: Add support to update placement policy for cloud zones in vRealize Automation

### PR DESCRIPTION
**Pull request by slakshmanan2706**
**Summary of Pull Request**

Enhances `Update-vRACloudAccountZone` and `Update-vRACloudZone` cmdlets to support placement policy for Cloud Zones in vRealize Automation

Supported placement policies: 
   - ADVANCED
   - DEFAULT 
   - SPREAD 
   - BINPACK
 
Adds new `Get-vRAAPIVersion` and `Get-vRAIntegrationDetail` cmdlets to support the same.

Signed-off-by: Lakshmanan Shanmugam <lshanmugam@vmware.com>

**Type of Pull Request**

- [ ] This is a bug fix.
- [x] This is an enhancement or feature.
- [ ] This is a code style / formatting update.
- [ ] This is a documentation update.
- [ ] This is a refactoring update.
- [ ] This is something else.
      Please describe:

**Context of the Pull Request***

**Related to Existing Issues**

Issue Number: 
Closes : #39 

**Test and Documentation Coverage**

- [x] Tests have been completed (for bug fixes / features).
- [ ] Documentation has been added / updated (for bug fixes / features).

**Breaking Changes?**

- [ ] Yes, there are breaking changes.
- [x] No, there are no breaking changes.

### CODE REVIEW

File: PowerValidatedSolutions.psm1 [hunk 1]
```diff
@@ -26350,24 +26426,46 @@ Function Update-vRACloudZone {
...
     #>
 
     Param (
         [Parameter (Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$id,
         [Parameter (ParameterSetName = "folder", Mandatory = $false)] [ValidateNotNullOrEmpty()] [String]$folder,
-        [Parameter (ParameterSetName = "tag", Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$tagKey,
-        [Parameter (ParameterSetName = "tag", Mandatory = $true)] [ValidateNotNullOrEmpty()] [String]$tagValue
+        [Parameter (ParameterSetName = "tag", Mandatory = $false)] [ValidateNotNullOrEmpty()] [String]$tagKey,
+        [Parameter (ParameterSetName = "tag", Mandatory = $false)] [ValidateNotNullOrEmpty()] [String]$tagValue,
+        [Parameter (ParameterSetName = "placementPolicy", Mandatory = $false)] [ValidateNotNullOrEmpty()] [String]$placementPolicy
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:26440]: Should this be a ValidateSet?

File: PowerValidatedSolutions.psm1 [hunk 2]
```diff
@@ -26350,24 +26426,46 @@ Function Update-vRACloudZone {
...
+   
     )
-
     Try {
         $cloudZoneDetails = Get-VRACloudZone -id $id
         if ($PsBoundParameters.ContainsKey("id") -and $PsBoundParameters.ContainsKey("folder")) {
-            $json = '{ "name": "' + $($cloudZoneDetails.name) + '", "folder": "' + $folder +'" }'
+            $json = '{ "name": "' + $($cloudZoneDetails.name) + '", "folder": "' + $folder + '" }'
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:26446]: Removed the intentional <space>.

File: PowerValidatedSolutions.psm1 [hunk 3]
```diff
@@ -26350,24 +26426,46 @@ Function Update-vRACloudZone {
...
+        }
+        if ($PsBoundParameters.ContainsKey("id") -and $PsBoundParameters.ContainsKey("placementPolicy")) {
+            if ($placementPolicy -eq "ADVANCED") {
+                if (((Get-vRAIntegrationDetail -integrationType vrops -getVCID).split(',')).contains(((Get-vRACloudAccount -type vsphere).customProperties.vcUuid))) {
+                    $cloudZoneDetails.placementPolicy = $placementPolicy
+                    $cloudZoneDetails.customProperties | Add-Member -Type NoteProperty  -Name "__ignoreAdvancedPolicyFailure" -Value "true" -Force
+                    $json = $cloudZoneDetails | ConvertTo-Json -Depth 4
+                }
+                else {
+                    Write-Error "vRealize Operations Manager is not integrated  with the vRealize Automation for the vCenter Server configured in CloudZone(id :$id) , check the integration: PRE_VALIDATION_FAILED"
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:26459]: "not integrated with vRealize Automation"

File: PowerValidatedSolutions.psm1 [hunk 4]
```diff
@@ -26335,6 +26348,69 @@ Function Get-vRACloudZone {
...
+    Try {
+        $vraapiVersion = "apiVersion=" + (Get-vRAAPIVersion)
+        $uri = "https://$vraAppliance/iaas/api/integrations?$vraapiVersion"        
+        $response = Invoke-RestMethod -Method Get -Uri $uri -Headers $vraHeaders
+        if ($getVCID) {
+            ($response.content | Where-Object integrationType -Match $integrationType).customProperties.vcIds
+        }
+        if ($getIntegrationID) {
+            if (($PsBoundParameters.ContainsKey("integrationType")) -and ($PsBoundParameters.ContainsKey("integrationName"))) {
+                (($response.content | Where-Object integrationType -Match $integrationType) | Where-Object  name -Match $integrationName).id
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:26404]: Extra `<space>` between `Where-Object` and `name`.

File: PowerValidatedSolutions.psm1 [hunk 5]
```diff
@@ -26335,6 +26348,69 @@ Function Get-vRACloudZone {
...
+    <#
+        .SYNOPSIS
+        Get an integration detail of an item from vRealize Automation
+
+        .DESCRIPTION
+        The Get-vRAIntegrationDetail cmdlet get an integration details of an item from vRealize Automation
+
+        .EXAMPLE
+        Get-vRAIntegrationDetail -integrationType "vrops" -getVCID
+        This example gets the vCenter Servers ids managed by the vRealize Operations Manager
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:26383]: "returns the ids of the vCenter Server instances"

File: PowerValidatedSolutions.psm1 [hunk 6]
```diff
@@ -26335,6 +26348,69 @@ Function Get-vRACloudZone {
...
+}
+Export-ModuleMember -Function Get-vRAAPIVersion
+
+Function Get-vRAIntegrationDetail {
+    <#
+        .SYNOPSIS
+        Get an integration detail of an item from vRealize Automation
+
+        .DESCRIPTION
+        The Get-vRAIntegrationDetail cmdlet get an integration details of an item from vRealize Automation
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:26379]: "gets" or "returns"

File: PowerValidatedSolutions.psm1 [hunk 7]
```diff
@@ -26335,6 +26348,69 @@ Function Get-vRACloudZone {
 }
 Export-ModuleMember -Function Get-vRACloudZone
 
+Function Get-vRAAPIVersion {
+    <#
+        .SYNOPSIS
+        Retrieve the vRealize Automation version information
+        
+        .DESCRIPTION
+        The GGet-vRAAPIVersion cmdlest retrieves the ApiVersion of vRealize Automation
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:26357]: "Get-" vs "GGet-"

Comment (tenthirtyam) [reply PowerValidatedSolutions.psm1:26357]: "cmdlet"

Comment (tenthirtyam) [reply PowerValidatedSolutions.psm1:26357]: returns the latest version of the API

File: PowerValidatedSolutions.psm1 [hunk 8]
```diff
@@ -13893,10 +13893,15 @@ Function Update-vRACloudAccountZone {
...
         - Adds a dynamic filter to use the defined tags
+        - Updates the placement policy
 
         .EXAMPLE
         Update-vRACloudAccountZone -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -domain sfo-w01 -vraUser configadmin -vraPass VMw@re1! -tagKey enabled -tagValue true -folder "sfo-w01-fd-workload" -resourcePool "sfo-w01-cl01-rp-workload"
         This example updates the Cloud Zone for the Workload Domain with a default folder and adds tags to the resource pool for dynamic provisioning in vRealize Automation
+        
+        .EXAMPLE
+        Update-vRACloudAccountZone -server sfo-vcf01.sfo.rainpole.io -user administrator@vsphere.local -pass VMw@re1! -domain sfo-w01 -vraUser configadmin -vraPass VMw@re1! -placementPolicy ADVANCED
+        This example updates the placement policy for Cloud Zones to ADVANCED in vRealize Automation
```
Comment (tenthirtyam) [inline PowerValidatedSolutions.psm1:13904]: "for the Cloud Zone"

**Review by tenthirtyam — CHANGES_REQUESTED** [review]
Added some comments for review and update - let me know if you have any questions.

File: PowerValidatedSolutions.psm1 (hunk 5 above)
Comment (slakshmanan2706) [reply PowerValidatedSolutions.psm1:26383]: Done

File: PowerValidatedSolutions.psm1 (hunk 6 above)
Comment (slakshmanan2706) [reply PowerValidatedSolutions.psm1:26379]: Done

File: PowerValidatedSolutions.psm1 (hunk 7 above)
Comment (slakshmanan2706) [reply PowerValidatedSolutions.psm1:26357]: Done

File: PowerValidatedSolutions.psm1 (hunk 8 above)
Comment (slakshmanan2706) [reply PowerValidatedSolutions.psm1:13904]: Done

**Review by tenthirtyam — APPROVED** [review]
Thanks for making the requests changes, @slakshmanan2706 - LGTM! 🚀.

**Ryan Johnson**
Senior Staff Solutions Architect | Product Engineering @ VMware, Inc.

**Pull request merged**